thrust/thrust/system/cuda/detail/copy.h

```diff
@@ -38,54 +38,178 @@
+  cuda_cub::throw_on_error(status, "trivial_device_copy D->H failed");
+}
+
+template <class H, class D, class InputIt, class Size, class OutputIt>
+OutputIt _CCCL_HOST non_trivial_cross_system_copy_n(
+  cpp::execution_policy<H>& host_s,
+  cuda_cub::execution_policy<D>& device_s,
+  InputIt first,
+  Size num_items,
+  OutputIt result)
+{
+  // copy input data into uninitialized host temp storage
+  using InputTy = thrust::detail::it_value_t<InputIt>;
+  thrust::detail::temporary_array<InputTy, H> temp_host(host_s, num_items);
+  // FIXME(bgruber): this fails to compile until #5490 is fixed
+  //::cuda::std::uninitialized_copy_n(first, num_items, temp_host.begin());
+  for (Size idx = 0; idx != num_items; idx++)
+  {
+    ::new (static_cast<void*>(temp_host.data().get() + idx)) InputTy(*first);
+    ++first;
+  }
+
+  // trivially copy data from host to device temp storage
+  thrust::detail::temporary_array<InputTy, D> temp_device(device_s, num_items);
+  const cudaError status = cuda_cub::trivial_copy_to_device(
+    temp_device.data().get(), temp_host.data().get(), num_items, cuda_cub::stream(device_s));
+  throw_on_error(status, "__copy:: H->D: failed");
+
+  // device->device copy
+  OutputIt ret = cuda_cub::copy_n(device_s, temp_device.data(), num_items, result);
+  return ret;
+}
 
-THRUST_NAMESPACE_BEGIN
-namespace cuda_cub
+#if _CCCL_HAS_CUDA_COMPILER()
+// non-trivial copy D->H, only supported with NVCC compiler
+// because copy ctor must have  __device__ annotations, which is nvcc-only
+// feature
+template <class D, class H, class InputIt, class Size, class OutputIt>
+OutputIt _CCCL_HOST non_trivial_cross_system_copy_n(
+  cuda_cub::execution_policy<D>& device_s,
+  cpp::execution_policy<H>& host_s,
+  InputIt first,
+  Size num_items,
+  OutputIt result)
 {
+  // copy input data into uninitialized device temp storage
+  using InputTy = thrust::detail::it_value_t<InputIt>;
+  thrust::detail::temporary_array<InputTy, D> temp_device(device_s, num_items);
+  cuda_cub::uninitialized_copy_n(device_s, first, num_items, temp_device.data());
+
+  // trivially copy data from device to host temp storage
+  thrust::detail::temporary_array<InputTy, H> temp_host(host_s, num_items);
+  const auto status = cuda_cub::trivial_copy_from_device(
+    temp_host.data().get(), temp_device.data().get(), num_items, cuda_cub::stream(device_s));
+  throw_on_error(status, "__copy:: D->H: failed");
+
+  // host->host copy
+  OutputIt ret = thrust::copy_n(host_s, temp_host.data(), num_items, result);
+  return ret;
+}
+#endif // _CCCL_HAS_CUDA_COMPILER()
+
+template <class System1, class System2, class InputIt, class Size, class OutputIt>
+OutputIt _CCCL_HOST cross_system_copy_n(cross_system<System1, System2> systems, InputIt begin, Size n, OutputIt result)
+{
+  if (n == 0)
+  {
+    return result;
+  }
+
+  // FIXME(bgruber): I think this is a pessimization. We should only check if the iterator is contiguous and the value
```

Comment:
I would argue that the other path is buggy rather than this being a pessimization